### PR TITLE
Turn the "no saved object file in work product" ICE into a translatable fatal error

### LINKED ISSUE #120854: Panic while compiling on 1.75.0, aarch64-apple-darwin
### Overview
I cannot share details, but I was doing `cargo test` and `cargo nextest` on my codebase after a `cargo sweep --time 10` and it had an error. Could this be due to incoherence of the cache?

### Meta
```
╰─➤  rustc --version --verbose
rustc 1.75.0 (82e1608df 2023-12-21)
binary: rustc
commit-hash: 82e1608dfa6e0b5569232559e3d385fea5a93112
commit-date: 2023-12-21
host: aarch64-apple-darwin
release: 1.75.0
LLVM version: 17.0.6
```

### Error output

```
Running cargo nextest
   Compiling my2_abc_types v0.1.0 (/Users/bob/work/3.myrepo/myrepo/lib/rust/my2/abc_types)
   Compiling customrouting_api v0.1.0 (/Users/bob/work/3.myrepo/myrepo/lib/rust/customrouting/api)
   Compiling my2_roleplay_model v0.1.0 (/Users/bob/work/3.myrepo/myrepo/lib/rust/my2/roleplay/model)
   Compiling repo_tools_bazel_build_codegen v0.1.0 (/Users/bob/work/3.myrepo/myrepo/lib/rust/repo_tools/bazel_build_codegen)
   Compiling asterisk_controller v0.1.0 (/Users/bob/work/3.myrepo/myrepo/services/asterisk_controller)
   Compiling generate_code v0.1.0 (/Users/bob/work/3.myrepo/myrepo/lib/rust/codegen/generate)
   Compiling commonlib_github v0.1.0 (/Users/bob/work/3.myrepo/myrepo/lib/rust/commonlib/github)
   Compiling goodb_storage_mysql v0.1.0 (/Users/bob/work/3.myrepo/myrepo/lib/rust/goodb/storage_mysql)
   Compiling goodb_api v0.1.0 (/Users/bob/work/3.myrepo/myrepo/lib/rust/goodb/api)
   Compiling my2_roleplay_api v0.1.0 (/Users/bob/work/3.myrepo/myrepo/lib/rust/my2/roleplay/api)
   Compiling customrouting_server v0.1.0 (/Users/bob/work/3.myrepo/myrepo/lib/rust/customrouting/server)
   Compiling playground v0.1.0 (/Users/bob/work/3.myrepo/myrepo/lib/rust/playground)
   Compiling my2_lib_content v0.1.0 (/Users/bob/work/3.myrepo/myrepo/lib/rust/my2/lib_shared/content)
thread 'cpy 20tg1ci7pxyvfa4j' panicked at /rustc/82e1608dfa6e0b5569232559e3d385fea5a93112/compiler/rustc_codegen_ssa/src/back/write.rs:917:45:
no saved object file in work product
stack backtrace:
thread 'cpy 2uyiqw3khmcm5oym' panicked at /rustc/82e1608dfa6e0b5569232559e3d385fea5a93112/compiler/rustc_codegen_ssa/src/back/write.rs:917:45:
no saved object file in work product
thread 'cpy 4i85ziw116l3ch2f' panicked at /rustc/82e1608dfa6e0b5569232559e3d385fea5a93112/compiler/rustc_codegen_ssa/src/back/write.rs:917:45:
no saved object file in work product
   0: _rust_begin_unwind
   1: core::panicking::panic_fmt
   2: core::option::expect_failed
   3: rustc_codegen_ssa::back::write::execute_copy_from_cache_work_item::<rustc_codegen_llvm::LlvmCodegenBackend>
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.

stack backtrace:
   0: _rust_begin_unwind
   1: core::panicking::panic_fmt
   2: core::option::expect_failed
   3: rustc_codegen_ssa::back::write::execute_copy_from_cache_work_item::<rustc_codegen_llvm::LlvmCodegenBackend>
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.

stack backtrace:
   0: _rust_begin_unwind
   1: core::panicking::panic_fmt
   2: core::option::expect_failed
   3: rustc_codegen_ssa::back::write::execute_copy_from_cache_work_item::<rustc_codegen_llvm::LlvmCodegenBackend>
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.

error: the compiler unexpectedly panicked. this is a bug.

error: the compiler unexpectedly panicked. this is a bug.

error: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust/issues/new?labels=C-bug%2C+I-ICE%2C+T-compiler&template=ice.md

note: rustc 1.75.0 (82e1608df 2023-12-21) running on aarch64-apple-darwin

note: we would appreciate a bug report: https://github.com/rust-lang/rust/issues/new?labels=C-bug%2C+I-ICE%2C+T-compiler&template=ice.md

note: rustc 1.75.0 (82e1608df 2023-12-21) running on aarch64-apple-darwin

note: compiler flags: --crate-type lib -C embed-bitcode=no -C debuginfo=2 -C split-debuginfo=unpacked -C incremental=[REDACTED]

note: some of the compiler flags provided by cargo are hidden

query stack during panic:
end of query stack
note: compiler flags: --crate-type lib -C embed-bitcode=no -C debuginfo=2 -C split-debuginfo=unpacked -C incremental=[REDACTED]

note: some of the compiler flags provided by cargo are hidden

query stack during panic:
end of query stack
note: we would appreciate a bug report: https://github.com/rust-lang/rust/issues/new?labels=C-bug%2C+I-ICE%2C+T-compiler&template=ice.md

note: rustc 1.75.0 (82e1608df 2023-12-21) running on aarch64-apple-darwin

note: compiler flags: --crate-type lib -C embed-bitcode=no -C debuginfo=2 -C split-debuginfo=unpacked -C incremental=[REDACTED]

note: some of the compiler flags provided by cargo are hidden

query stack during panic:
end of query stack
thread 'coordinator' panicked at compiler/rustc_middle/src/util/bug.rs:36:26:
/rustc/82e1608dfa6e0b5569232559e3d385fea5a93112/compiler/rustc_codegen_ssa/src/back/write.rs:1590:29: worker thread panicked
stack backtrace:
   0: std::panicking::begin_panic::<alloc::string::String>
   1: rustc_middle::util::bug::opt_span_bug_fmt::<rustc_span::span_encoding::Span>::{closure#0}
   2: rustc_middle::ty::context::tls::with_opt::<rustc_middle::util::bug::opt_span_bug_fmt<rustc_span::span_encoding::Span>::{closure#0}, !>::{closure#0}
   3: rustc_middle::ty::context::tls::with_context_opt::<rustc_middle::ty::context::tls::with_opt<rustc_middle::util::bug::opt_span_bug_fmt<rustc_span::span_encoding::Span>::{closure#0}, !>::{closure#0}, !>
   4: rustc_middle::util::bug::bug_fmt
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.

error: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust/issues/new?labels=C-bug%2C+I-ICE%2C+T-compiler&template=ice.md

note: rustc 1.75.0 (82e1608df 2023-12-21) running on aarch64-apple-darwin

note: compiler flags: --crate-type lib -C embed-bitcode=no -C debuginfo=2 -C split-debuginfo=unpacked -C incremental=[REDACTED]

note: some of the compiler flags provided by cargo are hidden

query stack during panic:
end of query stack
   Compiling goodb_coredb_server_side v0.1.0 (/Users/bob/work/3.myrepo/myrepo/lib/rust/goodb/coredb_server_side)
thread 'rustc' panicked at compiler/rustc_middle/src/util/bug.rs:36:26:
/rustc/82e1608dfa6e0b5569232559e3d385fea5a93112/compiler/rustc_codegen_ssa/src/back/write.rs:1964:17: panic during codegen/LLVM phase
stack backtrace:
   0: std::panicking::begin_panic::<alloc::string::String>
   1: rustc_middle::util::bug::opt_span_bug_fmt::<rustc_span::span_encoding::Span>::{closure#0}
   2: rustc_middle::ty::context::tls::with_opt::<rustc_middle::util::bug::opt_span_bug_fmt<rustc_span::span_encoding::Span>::{closure#0}, !>::{closure#0}
   3: rustc_middle::ty::context::tls::with_context_opt::<rustc_middle::ty::context::tls::with_opt<rustc_middle::util::bug::opt_span_bug_fmt<rustc_span::span_encoding::Span>::{closure#0}, !>::{closure#0}, !>
   4: rustc_middle::util::bug::bug_fmt
   5: <rustc_session::session::Session>::time::<rustc_codegen_ssa::back::write::CompiledModules, <rustc_codegen_ssa::back::write::OngoingCodegen<rustc_codegen_llvm::LlvmCodegenBackend>>::join::{closure#0}>
   6: <rustc_codegen_ssa::back::write::OngoingCodegen<rustc_codegen_llvm::LlvmCodegenBackend>>::join
   7: <rustc_codegen_llvm::LlvmCodegenBackend as rustc_codegen_ssa::traits::backend::CodegenBackend>::join_codegen
   8: <rustc_interface::queries::Linker>::link
   9: rustc_span::set_source_map::<core::result::Result<(), rustc_span::ErrorGuaranteed>, rustc_interface::interface::run_compiler<core::result::Result<(), rustc_span::ErrorGuaranteed>, rustc_driver_impl::run_compiler::{closure#1}>::{closure#0}::{closure#0}>
  10: <scoped_tls::ScopedKey<rustc_span::SessionGlobals>>::set::<rustc_interface::interface::run_compiler<core::result::Result<(), rustc_span::ErrorGuaranteed>, rustc_driver_impl::run_compiler::{closure#1}>::{closure#0}, core::result::Result<(), rustc_span::ErrorGuaranteed>>
  11: rustc_span::create_session_globals_then::<core::result::Result<(), rustc_span::ErrorGuaranteed>, rustc_interface::interface::run_compiler<core::result::Result<(), rustc_span::ErrorGuaranteed>, rustc_driver_impl::run_compiler::{closure#1}>::{closure#0}>
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.

error: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust/issues/new?labels=C-bug%2C+I-ICE%2C+T-compiler&template=ice.md

note: rustc 1.75.0 (82e1608df 2023-12-21) running on aarch64-apple-darwin

note: compiler flags: --crate-type lib -C embed-bitcode=no -C debuginfo=2 -C split-debuginfo=unpacked -C incremental=[REDACTED]

note: some of the compiler flags provided by cargo are hidden

query stack during panic:
end of query stack
error: could not compile `my_api` (lib)
warning: build failed, waiting for other jobs to finish...
error: command `/Users/bob/.rustup/toolchains/1.75.0-aarch64-apple-darwin/bin/cargo test --no-run --message-format json-render-diagnostics --workspace --exclude my-wasm --exclude my-android --target aarch64-apple-darwin` exited with code 101
```

</p>
</details>

### PR DESCRIPTION
**Pull request by saethlin**
I don't know if it's fair to say this fixes https://github.com/rust-lang/rust/issues/120854 but it surely makes the error reporting better and should encourage people with good instincts like @CinchBlue.

### CODE REVIEW

File: compiler/rustc_codegen_ssa/src/back/write.rs [hunk 1]
```diff
@@ -913,7 +913,9 @@ fn execute_copy_from_cache_work_item<B: ExtraBackendMethods>(
 
     let object = load_from_incr_comp_dir(
         cgcx.output_filenames.temp_path(OutputType::Object, Some(&module.name)),
-        module.source.saved_files.get("o").expect("no saved object file in work product"),
+        module.source.saved_files.get("o").unwrap_or_else(|| {
+            cgcx.create_dcx().emit_fatal(errors::NoSavedObjectFile { cgu_name: &module.name })
```
Comment (nnethercote) [inline compiler/rustc_codegen_ssa/src/back/write.rs:917]: Is "CGU" the rustc term for what LLVM calls a "module"? I'm trying to understand the `cgu_name`/`module.name` inconsistency here.

Comment (saethlin) [reply compiler/rustc_codegen_ssa/src/back/write.rs:917]: Yes, I think that's what's going on here. I decided to call this a CGU name based on how `CachedModuleCodegen` are created: https://github.com/rust-lang/rust/blob/b5c46dc5426038a49c95398bce30eeb20ec421e2/compiler/rustc_codegen_ssa/src/base.rs#L770-L773

Comment (nnethercote) [reply compiler/rustc_codegen_ssa/src/back/write.rs:917]: 👍

**Pull request merged**